factory.rb

```diff
@@ -0,0 +1,107 @@
+class Factory
+  def self.new(*attributes, keyword_init: false, &block)
+    flag = attributes[0].is_a?(String) && attributes[0].match(/[A-Z]/)
+    class_name = flag ? attributes.shift : nil
+    subclass = Class.new self do
+      class_eval(&block) if block_given?
+      send(:attr_accessor, *attributes)
+
+      class << self
+        define_method :new do |*args|
+          object = allocate
+          object.send(:initialize, *args)
+          object
+        end
+      end
+      define_method :initialize do |*values|
+        if !keyword_init
+          raise ArgumentError, 'Too many arguments' if values.size > attributes.size
+          raise ArgumentError, 'Wrong number of arguments' if values.size < attributes.size
+          values.each_with_index { |val, i| send("#{attributes[i]}=", val) }
+        elsif values[0].keys.sort != attributes.sort
+          redundant_args = values[0].keys.sort - attributes.sort
+          missing_args = attributes.sort - values[0].keys.sort
+          raise ArgumentError, "Do not match expected keys: #{missing_args}, #{redundant_args}"
+        else
+          values[0].map { |k, v| instance_variable_set "@#{k}", v }
+        end
+      end
+
+      define_method :[] do |member|
+        if member.is_a?(String) || member.is_a?(Symbol)
```

Comment:
Style/IfUnlessModifier: Favor modifier `if` usage when having a single-line body. Another good alternative is the usage of control flow `&&`/`||`.